### PR TITLE
Sbromberger/bfs improvements

### PR DESCRIPTION
**Pull request by sbromberger**
Starting with gdistance:

```
# this PR
julia> @benchmark LightGraphs.gdist2($g,1)
BenchmarkTools.Trial:
  memory estimate:  12.13 MiB
  allocs estimate:  11
  --------------
  minimum time:     182.685 ms (0.00% GC)
  median time:      189.283 ms (0.00% GC)
  mean time:        191.267 ms (0.00% GC)
  maximum time:     216.101 ms (0.00% GC)
  --------------
  samples:          26
  evals/sample:     1

# current master
julia> @benchmark LightGraphs.gdistances($g,1)
BenchmarkTools.Trial:
  memory estimate:  9.00 MiB
  allocs estimate:  6
  --------------
  minimum time:     228.320 ms (0.00% GC)
  median time:      238.889 ms (0.00% GC)
  mean time:        242.413 ms (0.57% GC)
  maximum time:     272.067 ms (0.00% GC)
  --------------
  samples:          21
  evals/sample:     1

# this PR
julia> @benchmark LightGraphs.gdist2($sg,1)
BenchmarkTools.Trial:
  memory estimate:  12.13 MiB
  allocs estimate:  11
  --------------
  minimum time:     126.385 ms (0.00% GC)
  median time:      135.065 ms (0.00% GC)
  mean time:        143.928 ms (0.00% GC)
  maximum time:     295.152 ms (0.00% GC)
  --------------
  samples:          35
  evals/sample:     1

# current master
julia> @benchmark LightGraphs.gdistances($sg,1)
BenchmarkTools.Trial:
  memory estimate:  9.00 MiB
  allocs estimate:  6
  --------------
  minimum time:     146.291 ms (0.00% GC)
  median time:      171.028 ms (0.00% GC)
  mean time:        177.256 ms (0.93% GC)
  maximum time:     302.972 ms (0.00% GC)
  --------------
  samples:          29
  evals/sample:     1
```

### CODE REVIEW

**Comment by sbromberger** [comment]
cc @jpfairbanks, @somil55, and/or @juliohm for a review.

**Comment by sbromberger** [comment]
I will merge this tomorrow unless there are any objections.

File: src/traversals/bfs.jl [hunk 1]
```diff
@@ -58,46 +58,48 @@ bfs_tree(g::AbstractGraph, s::Integer; dir=:out) = tree(bfs_parents(g, s; dir=di
...
+                    push!(next_level, i)
+                    vert_level[i] = n_level
+                    visited[i] = true
+                end
             end
         end
+        n_level += one(T)
+        empty!(cur_level)
+        cur_level, next_level = next_level, cur_level
+        sort!(cur_level)
```
Comment (jpfairbanks) [inline src/traversals/bfs.jl]: Does this sort help performance, or is it for consistency of the output?

File: src/traversals/bfs.jl [hunk 2]
```diff
@@ -58,46 +58,48 @@ bfs_tree(g::AbstractGraph, s::Integer; dir=:out) = tree(bfs_parents(g, s; dir=di
     gdistances!(g, source, dists)
 
 Fill `dists` with the geodesic distances of vertices in `g` from `source`.
-`dists` should be a vector of length `nv(g)`. Return `dists`.
-For vertices in disconnected components the default distance is -1.
```
Comment (jpfairbanks) [inline src/traversals/bfs.jl]: Is this for unsigned integer graph types? What about the edge case where `nv == typemax` and `g` is a path graph for example? This is more likely with smaller int types.

File: src/traversals/bfs.jl (hunk 1 above)
Comment (sbromberger) [reply src/traversals/bfs.jl]: It helps performance a good deal by maintaining cache locality. I did explicit tests.

File: src/traversals/bfs.jl (hunk 2 above)
Comment (sbromberger) [reply src/traversals/bfs.jl]: if `nv == typemax` then you're going to have massive problems across the board. The bottom line is that there has to be some upper sentinel. If you've got a 256-node graph and you're trying to shoehorn it into a `UInt8`, you're gonna have a bad time.

Comment (sbromberger) [reply src/traversals/bfs.jl]: (In fact, we should probably ensure that `squash` doesn't do that.)

**Pull request merged**